### PR TITLE
Fix restart point on mirror

### PR DESCRIPTION
**Pull request by xinzweb**
The test missing_xlog failed because the mirror got a long catch-up to do, since it never created a restart point.

The fix is to add restart point whenever a checkpoint record is replayed on the mirror.
We also patch up the missing_xlog to ensure a restart point is created before shutdown the mirror.

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -10271,9 +10271,15 @@ CreateRestartPoint(int flags)
 	 * IN_ARCHIVE_RECOVERY state and an older checkpoint, else do nothing;
 	 * this is a quick hack to make sure nothing really bad happens if
 	 * somehow we get here after the end-of-recovery checkpoint.
+	 *
+	 * FIXME: GPDB allow replay to also change the control file during
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: FIXME is not needed, its fixed :-)

File: src/test/walrep/expected/missing_xlog.out [hunk 2]
```diff
@@ -45,6 +45,9 @@ begin
 	end loop;
 end;
 $$ language plpgsql;
+-- checkpoint to ensure clean xlog replication before bring down mirror
+checkpoint;
+begin;end;
```
Comment (ashwinstar) [inline src/test/walrep/expected/missing_xlog.out]: Having being;end; only guarantees clog synced to mirror not replayed. Mostly will need check the gp_stat_replication and make sure replay clog location is fulfilled.

File: src/backend/access/transam/xlog.c (hunk 1 above)
Comment (xinzweb) [reply src/backend/access/transam/xlog.c]: Well, yes, we fix that part. I added another FIXME for task to merge the two modes.

File: src/test/walrep/expected/missing_xlog.out (hunk 2 above)
Comment (xinzweb) [reply src/test/walrep/expected/missing_xlog.out]: Yes, we added another function `wait_for_replication_replay` to ensure the mirror indeed caught up with the mirror on that checkpoint record.

File: src/test/walrep/expected/missing_xlog.out [hunk 3]
```diff
@@ -27,6 +27,33 @@ returns text as $$
...
+	i int;
+begin
+	i := 0;
+	loop
+		-- the condition is to check whether we have any mirror still catching up with primary.
+		-- 1. if mirror is still recovering (not streaming yet), both replay_location and sent_location
+		--    are NULL, so we add coalesce() to avoid false positive.
+		-- 2. otherwise, there will be streaming replication established between the primary and the mirror,
+		--    hence we just check the replay_location with sent_location.
+		if not exists (
```
[the remark this answers is not in the document]
Comment (xinzweb) [reply src/test/walrep/expected/missing_xlog.out]: nice push, yes, we can do it.

**Pull request merged**